internal/controllers/resourceslice/slicecleanup.go

```diff
@@ -0,0 +1,192 @@
+package resourceslice
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"time"
+
+	apiv1 "github.com/Azure/eno/api/v1"
+	"github.com/Azure/eno/internal/manager"
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// cleanupController is responsible for deleting resource slices when they are no longer needed by their composition.
+// It holds a finalizer on slices so they can't be deleted by the k8s GC controller until the composition has been deleted.
+// The controller has very little surface area: it deletes slices and removes their finalizers without modifying the composition.
+//
+// A non-cached apiserver client is used to be absolutely sure that a slice can be safely deleted.
+// But we avoid using it by excluding brand new resource slices and always checking the informer cache first.
+type cleanupController struct {
+	client        client.Client
+	noCacheReader client.Reader
+}
+
+func NewCleanupController(mgr ctrl.Manager) error {
+	c := &cleanupController{
+		client:        mgr.GetClient(),
+		noCacheReader: mgr.GetAPIReader(),
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&apiv1.ResourceSlice{}).
+		WatchesRawSource(source.Kind(mgr.GetCache(), &apiv1.Composition{}, c.newCompEventHandler())).
+		WithLogConstructor(manager.NewLogConstructor(mgr, "sliceCleanupController")).
+		Complete(c)
+}
+
+func (c *cleanupController) newCompEventHandler() handler.TypedEventHandler[*apiv1.Composition, reconcile.Request] {
+	fn := func(c *apiv1.Composition, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+		for _, syn := range []*apiv1.Synthesis{c.Status.CurrentSynthesis, c.Status.PreviousSynthesis} {
```

Comment:
Is there a scenario where we could miss something from inflight?

Reply:
Good question. I don't think so - the inflight synthesis never actually has any resource slices. They're added right before it's swapped into the current slot.

Reply:
But those slices _could_ still exist even if in flight never gets swapped (could that ever happen?), do we clean them up in that case?

Reply:
The same write that adds the resource slice refs also swaps the states, so it shouldn't be possible. That said- I could imagine this changing in the future so let's just add in-flight to the slice just in case.

Reply:
Could it be the case that slices are created but never refd? In any case, I think checking InFlight, should take care of it as no slice is ever not associated to a synthesis.

Reply:
It's possible they wouldn't be ref'd but in that case checking the in-flight synthesis won't actually enqueue anything. It's okay though since resource slice creation enqueues anyway.